### PR TITLE
Add complete type annotations to the public API

### PR DESCRIPTION
**Pull request by nicoddemus**
The main concern is the public API, so you will find many internal APIs with `Any` annotations: the internal API is quite frankly a mess, with many properties having multiple types through it (for example units and exponents often are passed around as a dict or as a list of tuples).

### CODE REVIEW

File: src/barril/units/unit_system_manager.py [hunk 1]
```diff
@@ -58,90 +64,86 @@ class UnitSystemIDError(KeyError):
...
 
 class UnitSystemManager(Singleton):
     """
     A service that manages the unit systems. It handles the unit system objects creation and
     selection.
 
-    :ivar list _object_refs:
-        List of weakref for all objects which represents a value+unit.
+    :ivar _object_refs:
+        Set of weakref for all objects which represents a value+unit.
```
Comment (tadeu) [inline src/barril/units/unit_system_manager.py:78]: Nice!

**Review by tadeu — APPROVED** [review]
Awesome improvement!

File: src/barril/units/_array.py [hunk 2]
```diff
@@ -31,45 +49,75 @@ class Array(AbstractValueWithQuantityObject):
         Array(ObtainQuantity("m", "length"), [0, 1, 2, 3, 4])
     """
 
-    def __init__(self, category, values=None, unit=None):
-        AbstractValueWithQuantityObject.__init__(self, category, value=values, unit=unit)
-
-    def _InternalCreateWithQuantity(self, quantity, values=None, unit_database=None, value=None):
+    @overload
```
Comment (oliveira-mauricio) [inline src/barril/units/_array.py:52]: That's a lot of options for init hehe.

File: src/barril/units/_quantity.py [hunk 3]
```diff
@@ -3,28 +3,67 @@
...
+def ObtainQuantity(
+    unit: Union[None, List[UnitExponentTuple], Dict[str, Sequence[Union[str, int]]]],
+    category: Optional[str] = None,
+    unknown_unit_caption: Optional[str] = None,
+) -> "Quantity":
+    ...
+
+
+def ObtainQuantity(
+    unit: Union[str, None, List[UnitExponentTuple], Dict[str, Sequence[Union[str, int]]]],
```
Comment (oliveira-mauricio) [inline src/barril/units/_quantity.py:63]: 👀

File: src/barril/units/_abstractvaluewithquantity.py [hunk 4]
```diff
@@ -33,11 +42,11 @@ def __init__(self, category, value=None, unit=None):
                 value = self._GetDefaultValue(quantity.GetCategoryInfo())
 
         else:
-            if category.__class__ != str:
+            if category.__class__ is not str:
```
Comment (oliveira-mauricio) [inline src/barril/units/_abstractvaluewithquantity.py:45]: Any reason why this one wasn't changed to `not isinstance(category , str)` ?

**Review by oliveira-mauricio — APPROVED** [review]
That's a lot of work, nice job.

File: src/barril/units/_quantity.py (hunk 3 above)
Comment (nicoddemus) [reply src/barril/units/_quantity.py:63]: Yep... that goes to show how type annotations actually *expose* hidden complexity.

File: src/barril/units/_abstractvaluewithquantity.py (hunk 4 above)
Comment (nicoddemus) [reply src/barril/units/_abstractvaluewithquantity.py:45]: Hmm no just oversight, fixed, thanks!

File: src/barril/units/_array.py (hunk 2 above)
Comment (nicoddemus) [reply src/barril/units/_array.py:49]: Yep, here overload really helps to unravel the hidden options we have.

File: src/barril/_util/types_.py [hunk 5]
```diff
@@ -27,20 +27,18 @@ def _GetKnownNumberTypes():
...
     :param object_:
         The object to check the type
 
-    :type type_: type, tuple of types, type name, tuple of type names
     :param type_:
         The type or types to check.
-        This can be a actual type or the name of the type.
-        This can be one type or a list of types.
 
     :param str message:
```
[the remark this answers is not in the document]
Comment (nicoddemus) [reply src/barril/_util/types_.py:43]: Yep missed those, thanks.

**Pull request merged**